New Shipping Apps section (#377)
crystal-lang/crystal-book: guides/shipping.md

```crystal
# staticserver.cr
require "http"
require "option_parser"

# Handle Ctrl+C and kill signal.
# Needed for hosting this process in a docker
# as the entry point command
Signal::INT.trap { puts "Caught Ctrl+C..."; exit }
Signal::TERM.trap { puts "Caught kill..."; exit }

path = "/www"
port = 80

option_parser = OptionParser.parse do |parser|
 parser.on "-f PATH", "--files=PATH", "Files path (default: /www)" do |files_path|
   path = files_path
 end
 parser.on "-p PORT", "--port=PORT", "Port to listen (default: 80)" do |server_port|
   port = server_port.to_i
 end
end

server = HTTP::Server.new([
 HTTP::LogHandler.new,
 HTTP::ErrorHandler.new,
 HTTP::StaticFileHandler.new(path),
])

address = server.bind_tcp "0.0.0.0", port
puts "Listening on http://#{address} and serving files in path #{path}"
server.listen
```

So, starting the server (listening on port 8080 and serving files under the current directory) is as easy as running:

```shell-session
$ crystal ./src/staticserver.cr -- -p 8080 -f .
Listening on http://0.0.0.0:8080 and serving files in path .
```

**Note:** the default behavior is to listen on `port 80` and serve the folder `/www`.

## Compiling our application

> Let’s go over Crystal’s [introduction](https://crystal-lang.org/reference/). One of the  main goals of the language is to _Compile to efficient native code_. That means that each time that we compile our code then an executable is built, but with an important property: it has a target platform (architecture and operating system), which is where the application will run. Crystal knows the target platform because is the same as the one being used to compile.

Reviewer: ```suggestion
Let’s go over Crystal’s [introduction](https://crystal-lang.org/reference/). One of the  main goals of the language is to _Compile to efficient native code_. That means that each time that we compile our code then an executable is built, but with an important property: it has a target platform (architecture and operating system), which is where the application will run. Crystal chooses the target platform by default as the same as the one being used to compile.
```

> For example, if we use a Linux OS based computer for compiling, then the executable will be meant to run on a Linux OS (and in some cases we will need to use the same Linux distribution).

Reviewer: `(and in some cases we will need to use the same Linux distribution)` - I'm not aware of any distribution-specific limitations. Do you have an example for that?

Reviewer: Sure, but that's not restricted by the distro, but the specific environment setup. You should be able to run a Linux executable on any Linux system (i.e. not needing to use the same distro), assuming the environment is configured appropriately.

> Can we set the target when calling the compiler? Oh, that’s a great idea, but for now it’s not an option (there are a lot of great buccanears [working on a solution](https://forum.crystal-lang.org/t/cross-compiling-automatically-to-osx/1330/12) and remember that Crystal is open source: so you are welcome aboard!)

Reviewer: This section suggest you can't cross compile Crystal. That is very much possible, though.

Only linking the final executable is not covered for a cross-compile target.

I'd either rephrase this paragraph or maybe even remove it entirely?

Reviewer 2: I agree to rephrase it. Cross compiling can be mention as something possible as long as you are aware of the difference between compiling, linking. But I would not aim for an explanation. At most that currently you would need a target environment to perform the linking.

Let’s compile our application:

```shell-session

> $ shards build --production

Reviewer: This is not about `shards`, so it rather only use `crystal build`.

Reviewer 2: I like that shards create a nicer directory structure.

But either we show the shards.yml and the structure of the whole folder.
Or we use `crystal build` (and update the samples below removing the `bin`).

But since we are simulating a whole app, using shards makes sense to me. Otherwise, it feels we are not using our own tools.
